File: fix_isar_ints.dart
import 'dart:io';

void main() {
  final targetDir = Directory('lib/data/local/isar_models');
  if (!targetDir.existsSync()) {
    print('Directory not found');
    return;
  }

  final files = targetDir.listSync().whereType<File>().where((file) => file.path.endsWith('.g.dart'));
  
  final maxSafeInteger = BigInt.parse('9007199254740991');
  final minSafeInteger = BigInt.parse('-9007199254740991');

  final pattern = RegExp(r'(?<![\w\.])(-?\d{16,})(?![\w\.])');

  for (final file in files) {
    final content = file.readAsStringSync();
    
    final newContent = content.replaceAllMapped(pattern, (match) {
      final numStr = match.group(1)!;
      final numBigInt = BigInt.tryParse(numStr);
      if (numBigInt != null && (numBigInt > maxSafeInteger || numBigInt < minSafeInteger)) {
        // Find JS safe integer
        final numDouble = double.tryParse(numStr);
        if (numDouble != null) {
          // dart:core double.toInt() might clamp to 64-bit integer, but we want the exact double representation
          // Double parsed from string will lose precision in exactly the same way JS does.
          // Wait, double in Dart web is the same as double in Dart native? Yes, IEEE 754 64-bit float.
          // So formatting it back to string as an exact integer (with zero decimal places)
          // `toStringAsFixed(0)` should give the rounded value.
          final rounded = numDouble.toStringAsFixed(0);
          return match.group(0)!.replaceFirst(numStr, rounded);
        }
      }
      return match.group(0)!;
    });

    if (content != newContent) {
      file.writeAsStringSync(newContent);
      print('Fixed ${file.path}');
    } else {
      print('No changes in ${file.path}');
    }
  }
}

File: fix_isar_ints.py
import os
import re
import glob

# Safe integer limit in JS
MAX_SAFE_INTEGER = 9007199254740991
MIN_SAFE_INTEGER = -9007199254740991

def replace_large_ints(match):
    num_str = match.group(1)
    num = int(num_str)
    if num > MAX_SAFE_INTEGER or num < MIN_SAFE_INTEGER:
        nearest = int(float(num))
        return match.group(0).replace(num_str, str(nearest))
    return match.group(0)

# We want to match `id: 1234567890123456789,` and other places where large integers might appear.
# Actually, let's just match any integer sequence that is 16 digits or more, 
# not preceded or followed by word characters or dots.
pattern = re.compile(r'(?<![\w\.])(-?\d{16,})(?![\w\.])')

target_dir = r"d:\Documments\Mobile code\Project_PRM\SmartFinance_Group3\lib\data\local\isar_models"
files = glob.glob(os.path.join(target_dir, "*.g.dart"))

for file_path in files:
    with open(file_path, "r", encoding="utf-8") as f:
        content = f.read()
    
    new_content = pattern.sub(replace_large_ints, content)
    
    if new_content != content:
        with open(file_path, "w", encoding="utf-8") as f:
            f.write(new_content)
        print(f"Fixed {os.path.basename(file_path)}")
    else:
        print(f"No changes in {os.path.basename(file_path)}")
